Update unittest for logger.py

Aijuh opened the pull request:
This closes #112 

I updated unittest for logger.py. 
Now, unittest and pylint has error at develop branch since I implemented unittest of old version of logger app and merge it.

This unittest does not have coverage rate 100% because of function `namer` inside `initLogger` in `LoggerApp`.
I don't have good idea about implementing unittest for function defined inside function.
So I am planning to make namer as private class method `_namer` if no idea is suggested.

Aijuh committed:
Update unittest for logger.py

I updated unittest for logger.py.

1. Add logFIlePath as argument for app.

2. Specify handler at setLevel.

giwon2004 commented on tests/test_logger.py:
```diff
@@ -20,10 +20,10 @@ def tearDown(self):
         with mock.patch("iquip.apps.logger._Signaller.signal") as mocked_signal:
-            app = logger.LoggerApp(name="name", parent=QObject())
-            app.setLevel("DEBUG")
+            app = logger.LoggerApp(name="name", logFilePath = "", parent=QObject())
```
We do not use whitespaces around `=` for keyword arguments.

There are more below, please change all of them.

Aijuh replied:
Thank you for review.
I missed it.

giwon2004 reviewed (changes requested):
Only one thing!

Aijuh committed:
Fix white space.

Aijuh commented:
Still waiting for the idea about implementing unittest of function defined inside function. I also searched for it, but cannot find valid reference. Otherwise I will make function `namer` as method of LoggerApp.

Unittest build 3.9 sometimes has following error

> qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

This error seems to come from `explore.py`.

BECATRUE commented:
> Unittest build 3.9 sometimes has following error

I don't know the specific reason, but sometimes the unittest may be failed though it is implemented correctly.
Then, you can just enter the details and re-run it!

![image](https://github.com/snu-quiqcl/iquip/assets/76851886/a1bb8551-d020-4d7d-b74d-1be3db5f0761)

BECATRUE commented on tests/test_logger.py:
```diff
@@ -67,51 +72,53 @@ def test_set_level(self):
     def test_call_check_to_clear(self):
         with mock.patch("iquip.apps.logger.LoggerApp.checkToClear") as mocked_method:
-            app = logger.LoggerApp(name="name", parent=QObject())
+            app = logger.LoggerApp(name="name", logFilePath="", parent=QObject())
```
Is it possible to move this statement outside of `with` block?

I think it looks better like `test_add_log()`.

Aijuh replied:
It is not possible because logger app's method is mocked.
When I move it outside of `with` block, the `checkToClear` method didn't called.
I think if app is created before `with` statement, app's `checkToClear` method is not same as mocked method.

BECATRUE replied:
Okay, then remain it as it is!

BECATRUE commented on tests/test_logger.py:
```diff
@@ -67,51 +72,53 @@ def test_set_level(self):
-            self.assertEqual(app.handler.level, level)
+            app.setLevel(app.frameHandler, levelText)
+            app.setLevel(app.fileHandler, levelText)
+            self.assertEqual(app.frameHandler.level, level)
```
Please check `app.fileHandler.level`, too!

BECATRUE commented on tests/test_logger.py:
```diff
@@ -67,51 +72,53 @@ def test_set_level(self):
-        self.assertEqual(prev_level, app.handler.level)
+        app.setLevel(app.frameHandler, non_level)
+        app.setLevel(app.fileHandler, non_level)
+        self.assertEqual(prev_level, app.frameHandler.level)
```
Check `fileHandler`!

BECATRUE reviewed (changes requested):
Umm.. Can you check `namer()` though checking `self.fileHandler.namer`?

For example,
```python
app.initLogger()

name = "blah"
modified_name = app.frameHandler.namer(name)

self.assertEqual(modified_name, "blah.log")
```

I'm not sure, but it seems to implement the unittest for `namer()`.

Aijuh committed:
check fileHandler & unittest for namer

Aijuh commented:
I added code for unittest of namer and code for checking level of fileHandler.
I implemented unittest of namer by following suggestion from @BECATRUE.
Thank you.

giwon2004 commented:
LGTM!

BECATRUE commented on tests/test_logger.py:
```diff
@@ -20,10 +20,10 @@ def tearDown(self):
         with mock.patch("iquip.apps.logger._Signaller.signal") as mocked_signal:
-            app = logger.LoggerApp(name="name", parent=QObject())
-            app.setLevel("DEBUG")
+            app = logger.LoggerApp(name="name", logFilePath="", parent=QObject())
```
It looks better to assign a non-empty string to `logFilePath`.
Is it possible?

Aijuh replied:
If I added `logFilepath = "test"`, it said directory `\\iquip\\test\\log_230821-144124.log'` is not exist.

BECATRUE replied:
Oh, I see. Then if we set `logFilePath` to an empty string, where is the log file created, in the iquip directory or the called directory?

BECATRUE replied:
Also, does the unittest make a log file? If so, the file should be removed after the test.

BECATRUE commented on tests/test_logger.py:
```diff
@@ -67,51 +72,61 @@ def test_set_level(self):
-                app = logger.LoggerApp(name="name", parent=QObject())
                 app.addLog("hello")
                 timeString = mocked_time().toString("yyyy-MM-dd HH:mm:ss")
                 msg = f"{timeString}: hello\n"
```
Oh, now I realize it.
Please add an empty line between this class and `if __name__ == "__main__":`

Aijuh committed:
Add empty line

Aijuh commented:
Good point. I should have checked that.
Since the test is running at the `iquip` directory, file is created at that directory with designated file name.
I will make code for removing that created dummy file.

kangz12345 reviewed (approved):
Just approving since I am a code owner (not reviewed).
Please mention me if you need my opinions.

Aijuh committed:
Remove dummy file for unittest

Add code for discard unwanted dummy file created by file handler.
First make directory "logfile-unittest" and create all the file inside that directory.
At the tearDown, discard connection to the created file and delete files and directory.

Aijuh committed:
add docstring

Aijuh commented:
I added codes for removing dummy file created by file handler.
In order to remove file, I should first discard connection to the open file.

I want to define that codes inside destructor `__del__` and call it by `del app`. But when I call `del app` and try to remove file, destructor does not called and the file is still opened. So I cannot remove dummy file.
As I understand, destructor is called immediately when reference count becomes zero. But as I think no one refer to the created app, but the desctuctor does not called. I cannot figure out way to use `__del__` method.
Therefore I made another method inside class `LoggerApp` in `logger.py`.

BECATRUE commented on tests/test_logger.py:
```diff
@@ -14,16 +15,24 @@ class LoggingHandlerTest(unittest.TestCase):
+        self.app.removeHandler()
+        file_list = os.listdir(self.logdir)
+        for file in file_list:
+            os.remove(self.logdir+"/"+file)
```
Please use `os.path.join()`.

BECATRUE commented on tests/test_logger.py:
```diff
@@ -14,16 +15,24 @@ class LoggingHandlerTest(unittest.TestCase):
 
     def setUp(self):
         self.qapp = QApplication([])
+        self.logdir = "logfile-unittest"
```
How about `log_dir`?

BECATRUE replied:
Please add a trailing "/" to show that the path is a directory explicitly.

BECATRUE replied:
Also, I think the directory path should be a full path!

Aijuh replied:
I thought that Unit test is executed at the location of `iquip`.
So I make another directory for dummy log file at the `iquip`, like `iquip/logfile-unittest` and delete that directory.
I don't know the exact directory where unittest is executed. 

Setting full path as `os.getcwd()+self.log_dir` would be same as previous implementation. Do you prefer this way? Or did I missed something?

BECATRUE replied:
As @Aijuh and I discussed yesterday, the current implementation doesn't seem to matter, so it is okay to remain it as it is.

Please just add a trailing "/"!

BECATRUE commented:
> I added codes for removing dummy file created by file handler. In order to remove file, I should first discard connection to the open file.
> 
> I want to define that codes inside destructor `__del__` and call it by `del app`. But when I call `del app` and try to remove file, destructor does not called and the file is still opened. So I cannot remove dummy file. As I understand, destructor is called immediately when reference count becomes zero. But as I think no one refer to the created app, but the desctuctor does not called. I cannot figure out way to use `__del__` method. Therefore I made another method inside class `LoggerApp` in `logger.py`.

Umm... We don't use `removeHandler()` while implementing, so I think it should be removed.
However, as you said, it looks a difficult problem.

Did you try overriding the destructor of `LoggerApp` and implementing the contents of `removeHandler()` in the customized destructor?

Aijuh committed:
Update file handling

Aijuh commented:
@BECATRUE 
`LoggerApp`'s parent class is `Baseapp` in qiwis.py, which does not have `__del__` method.
I tested for it but it does not work. I read that `del` method deletes class during garbage collection in python. 

There is way to make destructor as `__del__` and directly call `self.app.__del__()`. But is incurs pylint error.

BECATRUE commented:
> @BECATRUE `LoggerApp`'s parent class is `Baseapp` in qiwis.py, which does not have `__del__` method. I tested for it but it does not work. I read that `del` method deletes class during garbage collection in python.
> 
> There is way to make destructor as `__del__` and directly call `self.app.__del__()`. But is incurs pylint error.

@Aijuh 
I see. Then, how about moving `removeHandler()` into `test_logger.py`?
I still want to remove the method from `LoggerApp`.

Aijuh committed:
remove method from logger.py

Aijuh commented:
Sorry for the late update, @BECATRUE. 
Only `fileHandler.close()` is needed to close the open file.
So I removed the method at `logger.py` and add code `fileHandler.close()` at `test_logger.py`.

BECATRUE commented:
> Sorry for the late update, @BECATRUE. Only `fileHandler.close()` is needed to close the open file. So I removed the method at `logger.py` and add code `fileHandler.close()` at `test_logger.py`.

Wow, what a simple solution!

BECATRUE reviewed (approved):
LGTM!
